File: my-folder/1371-minimum-remove-to-make-valid-parentheses/solution.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        hs = set()
        stack = []

        for i in range(len(s)):
            c = s[i]

            if c == '(':
                stack.append(i)
            elif c == ')':
                if stack:
                    stack.pop()
                else:
                    hs.add(i)
            
        while stack:
            hs.add(stack.pop())

        res = ""

        for i in range(len(s)):
            if i not in hs:
                res += s[i]
        
        return res

File: my-folder/1371-minimum-remove-to-make-valid-parentheses/solution.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        Stack<Integer> stack = new Stack<>();
        Set<Integer> removeIndices = new HashSet<>();
        
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(') {
                stack.push(i);
            } else if (c == ')') {
                if (!stack.isEmpty()) {
                    stack.pop();
                } else {
                    removeIndices.add(i);
                }
            }
        }
        
        while (!stack.isEmpty()) {
            removeIndices.add(stack.pop());
        }
        
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            if (!removeIndices.contains(i)) {
                result.append(s.charAt(i));
            }
        }
        
        return result.toString();
    }
}
